internal/agentcfg/elasticsearch.go

```diff
@@ -0,0 +1,259 @@
+	}
+
+	if f.fallbackFetcher != nil {
+		f.metrics.fetchFallback.Add(1)
+		return f.fallbackFetcher.Fetch(ctx, query)
+	}
+
+	if f.invalidESCfg.Load() {
+		f.metrics.fetchFallbackUnavailable.Add(1)
+		f.logger.Errorf("rejecting fetch request: no valid elasticsearch config")
+		return Result{}, errors.New(ErrNoValidElasticsearchConfig)
+	}
+
+	f.metrics.fetchFallbackUnavailable.Add(1)
+	f.logger.Warnf("rejecting fetch request: infrastructure is not ready")
+	return Result{}, errors.New(ErrInfrastructureNotReady)
+}
+
+// Run refreshes the fetcher cache by querying Elasticsearch periodically.
+func (f *ElasticsearchFetcher) Run(ctx context.Context) error {
+	t := time.NewTicker(f.cacheDuration)
+	defer t.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-t.C:
+		}
+
+		if err := f.refreshCache(ctx); err != nil {
+			// Do not log as error when there is a fallback.
+			var logFunc func(string, ...interface{})
+			if f.fallbackFetcher == nil {
+				logFunc = f.logger.Errorf
+			} else {
+				logFunc = f.logger.Warnf
+			}
+
+			logFunc("refresh cache error: %s", err)
+			if f.invalidESCfg.Load() {
+				logFunc("stopping refresh cache background job: elasticsearch config is invalid")
+				return nil
+			}
+		} else {
+			f.logger.Debugf("refresh cache success")
+		}
+	}
+}
+
+func (f *ElasticsearchFetcher) refreshCache(ctx context.Context) (err error) {
+	scrollID := ""
+	buffer := make([]AgentConfig, 0, len(f.cache))
+
+	// The refresh cache operation should complete within refreshCacheTimeout.
+	ctx, cancel := context.WithTimeout(ctx, refreshCacheTimeout)
+	defer cancel()
+
+	defer func() {
+		if err != nil {
+			f.metrics.cacheRefreshFailures.Add(1)
+		} else {
+			f.metrics.cacheRefreshSuccesses.Add(1)
+		}
+	}()
+
+	var result struct {
+		Hits struct {
+			Hits []struct {
+				Source struct {
+					AgentName string `json:"agent_name"`
+					ETag      string `json:"etag"`
+					Service   struct {
+						Name        string `json:"name"`
+						Environment string `json:"environment"`
+					} `json:"service"`
+					Settings map[string]string `json:"settings"`
+				} `json:"_source"`
+			} `json:"hits"`
+		} `json:"hits"`
+		ScrollID string `json:"_scroll_id"`
+	}
+
+	for {
+		err = func() error {
+			var (
+				resp *esapi.Response
+				err  error
+			)
+			if scrollID == "" {
+				resp, err = esapi.SearchRequest{
+					Index:  []string{ElasticsearchIndexName},
+					Size:   &f.searchSize,
+					Scroll: f.cacheDuration,
+				}.Do(ctx, f.client)
+				if err != nil {
+					return err
+				}
+				defer resp.Body.Close()
+
+				if resp.StatusCode >= http.StatusBadRequest {
+					// Elasticsearch returns 401 on unauthorized requests and 403 on insufficient permission
+					if resp.StatusCode == http.StatusUnauthorized || resp.StatusCode == http.StatusForbidden {
+						f.invalidESCfg.Store(true)
+					}
+					return fmt.Errorf("refresh cache returns non-200 status: %d", resp.StatusCode)
```

Comment:
I'm thinking about changing this to
```
b, _ := io.ReadAll(resp.Body)
return fmt.Errorf("refresh cache elasticsearch query failed with status code %d: %s", resp.StatusCode, string(b))
```
and let the caller log it. But the downside is that it will still emit a potentially long log line at level WARN when `apm-server.agent.config.elasticsearch` is not configured and at level ERROR when it is configured.

Reply:
How about logging the status code on `error` level, and the full body on `debug` level?